File: src/lib/api/exam.ts
/**
 * AI Exam Generator API Client (Agent #7 Integrator)
 * ==================================================
 * Calls backend AI Exam Generator Agent endpoints to generate multi-format practice exams
 * and evaluate student submissions with automatic mastery scoring.
 */

export interface ExamQuestion {
  id: string;
  type: "mcq" | "short" | "long" | "numerical" | "conceptual";
  question: string;
  difficulty?: string;
  options?: Record<string, string>;
  correct_option?: string;
  model_solution?: string;
  max_marks?: number;
}

export interface PracticeExam {
  exam_id: string;
  title: string;
  topic: string;
  difficulty: string;
  total_marks: number;
  questions: ExamQuestion[];
  created_at: string;
}

export interface ExamEvaluationResponse {
  attempt_id: string;
  exam_id: string;
  score_percentage: number;
  earned_marks: number;
  total_marks: number;
  updated_mastery: number;
  question_feedback: Record<
    string,
    {
      question: string;
      type: string;
      user_answer: string;
      score: number;
      max_marks: number;
      is_correct: boolean;
      feedback: string;
    }
  >;
  planner_recommendation: string;
}

import { apiFetch, API_BASE_URL } from "./client";

export async function generatePracticeExam(
  studentId: string,
  topic?: string,
  difficulty = "medium",
  numQuestions = 5,
  courseId = "biol_101"
): Promise<PracticeExam> {
  const response = await apiFetch(`${API_BASE_URL}/ai/exam/generate`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      student_id: studentId || "demo_student",
      topic: topic || null,
      difficulty: difficulty,
      num_questions: numQuestions,
      course_id: courseId,
    }),
  });

  if (!response.ok) {
    throw new Error(`Exam generation failed with status ${response.status}`);
  }

  return await response.json();
}

export async function evaluatePracticeExam(
  examId: string,
  studentId: string,
  userAnswers: Record<string, string>
): Promise<ExamEvaluationResponse> {
  const response = await apiFetch(`${API_BASE_URL}/ai/exam/evaluate`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      exam_id: examId,
      student_id: studentId || "demo_student",
      user_answers: userAnswers,
    }),
  });

  if (!response.ok) {
    throw new Error(`Exam evaluation failed with status ${response.status}`);
  }

  return await response.json();
}

File: reference-backend/app/api/routes/exam.py
"""
AI Exam Generator Agent Routes (exam.py)
========================================

Exposes endpoints for generating practice exams across multi-format question types
(MCQs, Short, Long, Numerical, Conceptual) and evaluating submitted exam attempts.
"""

import logging
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session as DBSession

from app.api.deps import get_current_user, resolve_student_id
from app.db.session import get_db
from app.models.models import User
from app.schemas.schemas import (
    ExamGenerateRequest,
    ExamGenerateResponse,
    ExamEvaluateRequest,
    ExamEvaluateResponse,
)
from app.ai.agents.exam_agent import exam_agent

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/ai", tags=["Agent #7: AI Exam Generator Agent"])


@router.post("/exam/generate", response_model=ExamGenerateResponse, status_code=status.HTTP_200_OK)
async def generate_practice_exam(
    req: ExamGenerateRequest,
    db: DBSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> ExamGenerateResponse:
    """
    Generates a multi-format practice assessment tailored to student weak concepts and requested difficulty.
    """
    student_id = resolve_student_id(req.student_id, current_user)

    try:
        res = await exam_agent.generate_exam(
            db=db,
            student_id=student_id,
            topic=req.topic,
            difficulty=req.difficulty,
            num_questions=req.num_questions,
            course_id=req.course_id,
        )
        return ExamGenerateResponse(**res)
    except Exception as e:
        logger.error("Failed to generate exam: %s", str(e), exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"AI Exam Generator Agent error: {str(e)}",
        )


@router.post("/exam/evaluate", response_model=ExamEvaluateResponse, status_code=status.HTTP_200_OK)
async def evaluate_practice_exam(
    req: ExamEvaluateRequest,
    db: DBSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> ExamEvaluateResponse:
    """
    Evaluates a submitted practice exam, scores each question, updates Agent #2 Mastery, and returns recommendations.
    """
    student_id = resolve_student_id(req.student_id, current_user)

    try:
        res = await exam_agent.evaluate_exam(
            db=db,
            exam_id=req.exam_id,
            student_id=student_id,
            user_answers=req.user_answers,
        )
        return ExamEvaluateResponse(**res)
    except Exception as e:
        logger.error("Failed to evaluate exam: %s", str(e), exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"AI Exam Evaluation error: {str(e)}",
        )
